finality-aleph/src/data_io/data_provider.rs

```diff
@@ -296,3 +309,126 @@ impl<B: BlockT> aleph_bft::DataProvider<AlephData<B>> for DataProvider<B> {
         data
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        data_io::{
+            data_provider::{ChainTracker, ChainTrackerConfig},
+            AlephData, MAX_DATA_BRANCH_LEN,
+        },
+        testing::{client_chain_builder::ClientChainBuilder, mocks::aleph_data_from_blocks},
+        SessionBoundaries, SessionId, SessionPeriod,
+    };
+    use futures::channel::oneshot;
+    use std::future::Future;
+    use std::{sync::Arc, time::Duration};
+    use substrate_test_runtime_client::{
+        runtime::Block, DefaultTestClientBuilderExt, TestClientBuilder, TestClientBuilderExt,
+    };
+    use tokio::time::sleep;
+
+    const SESSION_LEN: u32 = 100;
+    // The lower the interval the less time the tests take, however setting this too low might cause
+    // the tests to fail. Even though 1ms works with no issues, we set it to 5ms for safety.
+    const REFRESH_INTERVAL_MILLIS: u64 = 5;
+
+    fn prepare_chain_tracker_test() -> (
+        impl Future<Output = ()>,
+        oneshot::Sender<()>,
+        ClientChainBuilder,
+        impl aleph_bft::DataProvider<AlephData<Block>>,
+    ) {
+        let (client, select_chain) = TestClientBuilder::new().build_with_longest_chain();
+        let client = Arc::new(client);
+
+        let chain_builder =
+            ClientChainBuilder::new(client.clone(), Arc::new(TestClientBuilder::new().build()));
+        let session_boundaries = SessionBoundaries::new(SessionId(0), SessionPeriod(SESSION_LEN));
+
+        let config = ChainTrackerConfig {
+            refresh_interval: Duration::from_millis(REFRESH_INTERVAL_MILLIS),
+        };
+
+        let (chain_tracker, data_provider) =
+            ChainTracker::new(select_chain, client, session_boundaries, config, None);
+
+        let (exit_chain_tracker_tx, exit_chain_tracker_rx) = oneshot::channel();
+        (
+            async move {
+                chain_tracker.run(exit_chain_tracker_rx).await;
+            },
+            exit_chain_tracker_tx,
+            chain_builder,
+            data_provider,
+        )
+    }
+
+    // Sleep enough time so that the internal refreshing in ChainTracker has time to finish.
+    async fn sleep_enough() {
+        sleep(Duration::from_millis(2 * REFRESH_INTERVAL_MILLIS)).await;
+    }
+
+    async fn run_test<F, S>(scenario: S)
+    where
+        F: Future,
+        S: FnOnce(ClientChainBuilder, Box<dyn aleph_bft::DataProvider<AlephData<Block>>>) -> F,
+    {
+        let (task_handle, exit, chain_builder, data_provider) = prepare_chain_tracker_test();
+        let chain_tracker_handle = tokio::spawn(task_handle);
+
+        scenario(chain_builder, Box::new(data_provider)).await;
+
+        exit.send(()).unwrap();
+        chain_tracker_handle.await.unwrap();
```

Comment:
nit: unwrap -> expect for easier debug in future

Reply:
Ok